lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee

```diff
@@ -21,30 +21,47 @@ class TwitterCldr.NumberFormatter
     @constructor.data().symbols
 
   format: (number, options = {}) ->
-    opts = this.default_format_options_for(number)
+    opts = @default_format_options_for(number)
 
     for key, val of options
       opts[key] = if options[key]? then options[key] else opts[key]
 
-    tokens = this.get_tokens(number, opts)
+    number = @round_to(number, opts.precision || 0)
+
+    tokens = @get_tokens(number, opts)
 
     if tokens?
       if tokens not instanceof Array
         tokens_sample = tokens[Object.keys(tokens)[0]]
         truncated_number = @truncate_number(number, tokens_sample[1].length)
+
+        rounded_num = Number(@parse_number(truncated_number, opts)[0]) * Number(@get_key(number)) / (10 ** (tokens_sample[1].length - 1))
+
+        # Checking _stack_depth should be a redundant condition, as
+        # @get_key(rounded_num) should always equal @get_key(number) the second
+        # time around. We only check it to prevent infinite loops in case of
+        # bugs
+        if !opts._stack_depth && @get_key(rounded_num) != @get_key(number)
+          return @format(rounded_num, Object.assign(opts, { _stack_depth: 1 }))
+
         truncated_number = Math.floor(truncated_number) if opts.precision == 0
         tokens = tokens[TwitterCldr.PluralRules.rule_for(truncated_number)]
 
-      [prefix, suffix, integer_format, fraction_format] = this.partition_tokens(tokens)
-      number = this.truncate_number(number, integer_format.format.length)
-      [intg, fraction] = this.parse_number(number, opts)
+      [prefix, suffix, integer_format, fraction_format] = @partition_tokens(tokens)
+      number_part = @truncate_number(number, integer_format.format.length)
+      [intg, fraction] = @parse_number(number_part, opts)
       result = integer_format.apply(parseFloat(intg), opts)
       result += fraction_format.apply(fraction, opts) if fraction
-      sign = if number < 0 && prefix != "-" then @symbols().minus_sign || @default_symbols.minus_sign else ""
+      sign = if number_part < 0 && prefix != "-" then @symbols().minus_sign || @default_symbols.minus_sign else ""
       "#{prefix}#{result}#{suffix}"
     else
-      # there's no specific formatting pattern for this number in current locale
-      number.toString()
+      unless @ instanceof TwitterCldr.DecimalFormatter
+        # No specific formatting pattern for this number in current locale, so
+        # we fall back to DecimalFormatter
+        new TwitterCldr.DecimalFormatter().format(number, opts)
```

Comment:
Tbh I don't love instantiating the derived class from the base class, but for derived classes other than `DecimalFormatter`, it seems like a more sensible default than just calling `toString` (which completely ignores the correct formatting for that locale and just formats using `.` as a decimal point). Maybe there's a less kludgey way of refactoring this, though?

Reply:
Is this necessary as a fallback in case there's no formatting data for the language/number combo?

Reply:
TBH I've already forgotten where this was necessary — it seems the parent `else` code path should never actually be hit with any of the current `NumberFormatter`-derived classes, as `tokens` should always be defined. In which case could remove the `unless` and keep the fallback as `toString`, with a comment noting that it's only to prevent errors in case of bugs?